File: solutions/Virtual-Assistant/src/csharp/assistant/Dialogs/Onboarding/OnboardingDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using VirtualAssistant.Dialogs.Shared;

namespace VirtualAssistant.Dialogs.Onboarding
{
    public class OnboardingDialog : EnterpriseDialog
    {
        // Constants
        public const string NamePrompt = "namePrompt";

        // Fields
        private static OnboardingResponses _responder = new OnboardingResponses();
        private IStatePropertyAccessor<OnboardingState> _accessor;
        private OnboardingState _state;

        public OnboardingDialog(BotServices botServices, IStatePropertyAccessor<OnboardingState> accessor, IBotTelemetryClient telemetryClient)
            : base(botServices, nameof(OnboardingDialog), telemetryClient)
        {
            _accessor = accessor;
            InitialDialogId = nameof(OnboardingDialog);

            var onboarding = new WaterfallStep[]
            {
                AskForName,
                FinishOnboardingDialog,
            };

            // To capture built-in waterfall dialog telemetry, set the telemetry client
            // to the new waterfall dialog and add it to the component dialog
            TelemetryClient = telemetryClient;
            AddDialog(new WaterfallDialog(InitialDialogId, onboarding) { TelemetryClient = telemetryClient });
            AddDialog(new TextPrompt(NamePrompt));
        }

        public async Task<DialogTurnResult> AskForName(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            return await sc.PromptAsync(NamePrompt, new PromptOptions()
            {
                Prompt = await _responder.RenderTemplate(sc.Context, "en", OnboardingResponses.ResponseIds.NamePrompt),
            });
        }

        public async Task<DialogTurnResult> FinishOnboardingDialog(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());
            _state.Name = (string)sc.Result;

            await _responder.ReplyWith(sc.Context, OnboardingResponses.ResponseIds.Greeting, new { _state.Name });

            return await sc.EndDialogAsync();
        }
    }
}

File: solutions/Virtual-Assistant/src/typescript/assistant/src/dialogs/onboarding/onboardingDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import {
    BotTelemetryClient,
    StatePropertyAccessor,
    TurnContext } from 'botbuilder';
import {
    DialogTurnResult,
    TextPrompt,
    WaterfallDialog,
    WaterfallStep,
    WaterfallStepContext} from 'botbuilder-dialogs';
import { BotServices } from '../../botServices';
import { EnterpriseDialog } from '../shared/enterpriseDialog';
import { OnboardingResponses } from './onboardingResponses';
import { IOnboardingState } from './onboardingState';

export class OnboardingDialog extends EnterpriseDialog {
    // Fields
    private static readonly responder: OnboardingResponses = new OnboardingResponses();
    private accessor: StatePropertyAccessor<IOnboardingState>;
    private state!: IOnboardingState;

    // Constructor
    constructor(botServices: BotServices, accessor: StatePropertyAccessor<IOnboardingState>, telemetryClient: BotTelemetryClient) {
        super(botServices, OnboardingDialog.name, telemetryClient);
        this.accessor = accessor;
        this.initialDialogId = OnboardingDialog.name;
        const onboarding: ((sc: WaterfallStepContext<{}>) => Promise<DialogTurnResult>)[] = [
            this.askForName.bind(this),
            this.askForLocation.bind(this),
            this.finishOnboardingDialog.bind(this)
        ];

        this.addDialog(new WaterfallDialog(this.initialDialogId, onboarding));
        this.addDialog(new TextPrompt(DialogIds.namePrompt));
        this.addDialog(new TextPrompt(DialogIds.locationPrompt));
    }

    private async askForName(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        return sc.prompt(DialogIds.namePrompt, {
            prompt: await OnboardingDialog.responder.renderTemplate
            (
                sc.context,
                OnboardingResponses.responseIds.namePrompt,
                'en'
            )
        });
    }

    private async askForLocation(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        this.state = await this.getStateFromAccessor(sc.context);
        this.state.name = <string>sc.result;

        return sc.prompt(DialogIds.locationPrompt, {
            prompt: await OnboardingDialog.responder.renderTemplate
            (
                sc.context,
                OnboardingResponses.responseIds.locationPrompt,
                'en',
                this.state.name
            )
        });
    }

    private async finishOnboardingDialog(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        this.state = await this.getStateFromAccessor(sc.context);
        this.state.location = <string>sc.result;
        await OnboardingDialog.responder.replyWith(
            sc.context,
            OnboardingResponses.responseIds.haveLocation,
            this.state.location
        );
        await OnboardingDialog.responder.replyWith(
            sc.context,
            OnboardingResponses.responseIds.addLinkedAccountsMessage
        );

        return sc.endDialog();
    }

    private async getStateFromAccessor(context: TurnContext): Promise<IOnboardingState>  {
        const state: IOnboardingState | undefined = await this.accessor.get(context);
        if (!state) {
            const newState: IOnboardingState = {
                dialogStack: [],
                location: '',
                name: ''
            };
            await this.accessor.set(context, newState);

            return newState;
        }

        return state;
    }
}

enum DialogIds {
    namePrompt = 'namePrompt',
    locationPrompt = 'locationPrompt'
}
